### DISCUSSION TITLE
Problem with "inconsistent sorting"

### DISCUSSION THREAD

**Discussion by javafanboy** [duckdb/duckdb#14915] [Q&A] [answered]
As part of some analytic workflow I do among other things analyze som logs. A small part (selection of columns and rows) looks like this (the num column is only added for debugging purposes, does not exist in the real dataset and cant be added with acceptable performance using the current data pipeline):

```
tier,host,threadId,timeMillis,timeNanos,num
tier,MAC-QXD7D57Q,38,1730627090244,666000,99991
tier,MAC-QXD7D57Q,38,1730627090244,666000,99992
tier,MAC-QXD7D57Q,38,1730627090244,667000,99993
tier,MAC-QXD7D57Q,38,1730627090244,667000,99994
tier,MAC-QXD7D57Q,38,1730627090244,668000,99995
tier,MAC-QXD7D57Q,38,1730627090244,714000,99996
tier,MAC-QXD7D57Q,38,1730627090244,715000,99997
tier,MAC-QXD7D57Q,38,1730627090244,716000,99998
tier,MAC-QXD7D57Q,38,1730627090244,717000,99999
```

The tables "natural ordering" is as shown, ie sorteds in ASC order on the columns in the order from left to right first tier, then host etc. making searching on a "composite key" of these columns very fast with DuckDB. The timeNanos field holds nanos within the millisecond and due to clock resolution and other quirks in logging systems it rarely provide full nano resolution and more likely at most microseconds within the millisecond resulting in possible duplicates of the sort keys. 

When I perform a query (once again simplified to show the relevant aspects) to extract different "pages" of the data. Paging only using OFFSET and LIMIT turned out to be too slow so I am leveraging information about previous "pages" and get the next page relative to that. The OFFSET will not typically be zero (except if the same page is requested twice in a row and removing it makes no difference for this problem). This is the query used when the new "page" is located earlier in the table than the previous one:

```
SELECT * FROM (SELECT tier,host,threadId,timeMillis,timeNanos,num FROM events WHERE (tier,host,threadId,timeMillis,timeNanos) <= ('tier', 'MAC-QXD7D57Q', 38, 1730627090244, 668000) ORDER BY tier DESC, host DESC, threadId DESC, timeMillis DESC, timeNanos DESC OFFSET 0 LIMIT 6) ORDER BY tier ASC, host ASC, threadId ASC, timeMillis ASC, timeNanos ASC;
tier,host,threadId,timeMillis,timeNanos,num
tier,MAC-QXD7D57Q,38,1730627090244,666000,99990
tier,MAC-QXD7D57Q,38,1730627090244,666000,99991
tier,MAC-QXD7D57Q,38,1730627090244,666000,99992
tier,MAC-QXD7D57Q,38,1730627090244,667000,99993
tier,MAC-QXD7D57Q,38,1730627090244,667000,99994
tier,MAC-QXD7D57Q,38,1730627090244,668000,99995

```
But with a one step smaller LIMIT the query gives an unexpected result (the row with "num" 99992 is missing) :
```
SELECT * FROM (SELECT tier,host,threadId,timeMillis,timeNanos,num FROM events WHERE (tier,host,threadId,timeMillis,timeNanos) <= ('tier', 'MAC-QXD7D57Q', 38, 1730627090244, 668000) ORDER BY tier DESC, host DESC, threadId DESC, timeMillis DESC, timeNanos DESC OFFSET 0 LIMIT 5) ORDER BY tier ASC, host ASC, threadId ASC, timeMillis ASC, timeNanos ASC;
tier,host,threadId,timeMillis,timeNanos,num
tier,MAC-QXD7D57Q,38,1730627090244,666000,99990
tier,MAC-QXD7D57Q,38,1730627090244,666000,99991
tier,MAC-QXD7D57Q,38,1730627090244,667000,99993
tier,MAC-QXD7D57Q,38,1730627090244,667000,99994
tier,MAC-QXD7D57Q,38,1730627090244,668000,99995

```
My guess is that the "root cause" is the fact that the ordering criteria result in duplicates that is handled differently in the "ORDER BY" of the nested query and the main query.

Is this intended behaviour with DuckDB and if so what would be the least inefficient way to rewrite my query to get the expected result? My where clause may (before limit is applied) still leave several million lines so I would like to avoid having to introduce a synthetic ROWNUM column etc. in the inner SELECT - something I otherwise would guess could help...

I am using Mac M3 and DuckDB 1.1.3.

All sugestions are much appreciated!

**Comment by javafanboy** [comment]
After further investigation I noticed that even without a nested query I get the same unexpected result (i.e. row 99992 is missing):
```
SELECT tier,host,threadId,timeMillis,timeNanos,num FROM events WHERE (tier,host,threadId,timeMillis,timeNanos) <= ('tier', 'MAC-QXD7D57Q', 38, 1730627090244, 668000) ORDER BY tier DESC, host DESC, threadId DESC, timeMillis DESC, timeNanos DESC OFFSET 0 LIMIT 5;
tier,host,threadId,timeMillis,timeNanos,num
tier,MAC-QXD7D57Q,38,1730627090244,668000,99995
tier,MAC-QXD7D57Q,38,1730627090244,667000,99993
tier,MAC-QXD7D57Q,38,1730627090244,667000,99994
tier,MAC-QXD7D57Q,38,1730627090244,666000,99990
tier,MAC-QXD7D57Q,38,1730627090244,666000,99991
```

so now I am even more confused but as before all help on why it behaves like this and of course how to solve it is very appreciated!

**Comment by javafanboy** [comment]
If it makes any difference the full schema for the table (with the added num field for debugging - not present in the real table) is as follows:

```
describe table events;
column_name,column_type,null,key,default,extra
tier,VARCHAR,YES,,,
host,VARCHAR,YES,,,
ip,VARCHAR,YES,,,
threadId,BIGINT,YES,,,
timeMillis,BIGINT,YES,,,
timeNanos,BIGINT,YES,,,
loggerName,VARCHAR,YES,,,
level,SMALLINT,YES,,,
message,VARCHAR,YES,,,
messageParameters,VARCHAR[],YES,,,
templateDigest,BIGINT,YES,,,
marker,VARCHAR,YES,,,
thrownParameters,VARCHAR[],YES,,,
className,VARCHAR,YES,,,
methodName,VARCHAR,YES,,,
lineNumber,INTEGER,YES,,,
contextData,"MAP(VARCHAR, VARCHAR)",YES,,,
contextStack,VARCHAR[],YES,,,
threadPriority,TINYINT,YES,,,
threadName,VARCHAR,YES,,,
rownum,INTEGER,YES,,,
num,BIGINT,YES,,,
```

**Comment by javafanboy** [comment]
I am guessing this could be due to non-stable sorting of equal rows in DuckDB - but this article https://duckdb.org/2021/08/27/external-sorting.html states DuckDB uses some variation of Radix Sort that normally is stable...
Anybody that know more about this or have some other sugestions about this problem?

**Comment by soerenwolfers** [comment]
Please format your code using triple back ticks. Also, to simplify the life of those wanting to help you, please boil down your issues to more minimal examples that focus on the problem at hand rather than irrelevant details. 

That being said, this does look like a bug, so please report this as an issue once you have a reproducible MWE, i.e. an example that generates the underlying data in SQL (e.g. using a ` VALUES` clause or `range` queries) or includes a link to a CSV or parquet file.

**Reply by javafanboy** [reply]
Thanks for suggestions - how much "irrelevant details" to provide is always hard and seem to dpeend from forum to forum - when not providing background in many forums I typically get a lot of comments from people critically asking why I ask this question / need to do what I need to do etc. and in order to avoid that I rather write too much than too little these days :-)

> Please format your code using triple back ticks. Also, to simplify the life of those wanting to help you, please boil down your issues to more minimal examples that focus on the problem at hand rather than irrelevant details.
> 
> That being said, this does look like a bug, so please report this as an issue once you have a reproducible MWE, i.e. an example that generates the underlying data in SQL (e.g. using a ` VALUES` clause or `range` queries) or includes a link to a CSV or parquet file.

**Reply by soerenwolfers** [reply]
I feel your stackoverflow trauma ;) I guess the difference there is that questions are usually of the sort "how do I best approach the problem" whereas here (at least in this case) it's "is there a bug with duckdb" and to verify that it helps if the user removes as much noise as possible without removing the potentially buggy behavior.

**Reply by javafanboy** [reply]
TBH I was not too sure if this was a bug or not as I am not an expert of what the SQL specification requires - perhaps non stable sorting is totally acceptable and the way a kind of specialized project like DuckDB could go too gain speed / parallelization?!

I will put together some testable example code and post as an "issue" and we see what comes of it...

**Reply by soerenwolfers** [reply]
Why do you think this has anything to do with stability? Stability concerns the treatment of rows that are equal in the given sort keys. The row you're missing is not equal to any other rows in the sort keys you're specifying, so its place in the order is independent of stability.

**Comment by javafanboy** [comment]
As I mentioned in the post the LAST column "num" is only added for writing this post and debugging the problem and is not present in the real dataset, not part of the natural sort order of the tabler or the ordering statements in the queries. This means that the rows with number 99990, 99991, 99992 have the same composite key and when I reduce the "window" from 6 to 5, if not respecting the native sort order, DuckDB could skip any of those three rows (for instance the missing 99992). If on the other hand sorting was stable I do not see how that line could be excluded....

**Reply by soerenwolfers** [reply, marked as answer]
ah sorry in that case, yeah, it's a result of non-stable sorting. you might get the stability you hope for (or at least determinism) by turning off multi threading using `SET threads=1`

**Comment by javafanboy** [comment]
Is it documented somewhere ift DuckDB indeed uses non-stable sorting? This
is after all an important thing to know about any sorting algorithm.

 If using say merge sort it is possible to do multi-threading and still
keep stability.

I will try that switch and see if it makes any difference even though
running with only a single thread is not an option for my use-case...

**Reply by javafanboy** [reply]
I did some simple tests that indeed confirms that DuckDB sorting is unstable. Personally I think this is the wrong tradeoff as there are fast stable methods available. In analytical processing it is by no means uncommon with dataset that do not have unique keys (similar to mine with duplicated date/time fields for instance) but it is what it is....
